### DISCUSSION TITLE
Model interfaces with shared cabling

### DISCUSSION THREAD

**Discussion by nward** [nautobot/nautobot#587] [Ideas]
### Environment
* Python version: any
* Nautobot version: future

### Proposed Functionality

- Add modelling for multiple interfaces which share a single physical cable - such as QSFP+ going to a breakout cable or panel.
- Add modelling for cables with different numbers of connectors at each end (i.e. MPO one end, 4 x LC pair the other), and a way to create new cables / pinouts to describe different cable types (MPO-MPO A/B/C polarity, MPO-4xLC etc.)

### Use Case

PSM4(PLR4) optics have an MPO/MTP connector, and are recognised by the device a 4 separate physical interfaces (and can be 10/25/100g per interface).

There are 2 ways these are used:
- MPO breakout cable, which has an MPO connector on one end, and 4 x LC pairs on the other.
- MPO breakout panel, usually a cassette which fits in a panel, and has 4 x LC pairs on the front, and 1 x MPO on the back. An MPO cable is connected from the panel to the QSFP+ in the equipment.

Multimode QSFP+ optics can be used with a breakout MPO cable as in the same way.

Currently, we have to model the 4 physical interfaces as SFP+ 10G, and connect cables directly to them - and cannot model the existence of either a breakout cable, or a breakout panel.

Some users achieve this by creating front/back connections in the device with the QSFP+ optic installed - however this is very cumbersome and does not accurately represent reality.

### Database Changes

I'm not sure - proposing this so we can discuss the requirements, then address how to achieve this in the DB.

I think fibre modelling should be changed, to model individual strands rather than pairs - trying to build this feature without this is possible, but means that there needs to be code for all the different options - better do model fibre strands so users can define cables / etc. themselves.

I expect cables would need to be modelled differently to allow different numbers of connectors.

### External Dependencies

I expect no external dependency changes.

**Comment by jathanism** [comment]
Thanks @nward for the request. We'll give this a good review and follow up.

**Comment by jedelman8** [comment]
We're totally open to this, but definitely need to discuss requirements and design proposals.

**Comment by jedelman8** [comment]
@nward You may have seen the issue be closed, but just a heads up converted this to a discussion to get more input and feedback.

**Comment by stefan-baranoff** [comment]
I've been thinking about this problem significantly recently. I have a model that works, but is annoying as heck to build out today and isn't quite "right". In my use cases "shared cabling" is in terms of breakout DACs/fiber. In this case the model I'm using is:
 - Interfaces are the logical broken out interfaces the switch config deals with
 - These are "cabled" 1 to 1 to front ports labeled "inside of transceiver"
 - The "inside of transceiver" front ports are mapped in groups of 4 to a 4 position rear ports labeled "outside of transceiver"
 - The "outside of transceiver" rear port is cabled to another 4 position rear port labeled "physical port"
 - The "physical port" rear port is mapped to 4 front LC ports which are then cabled to the far devices

This ALMOST models reality. It's sufficient for building the logical configuration on the device AND port/cable planning. The only "got you" in it is that there are many "cables" that don't actually exist and are either internal traces/ICs (in the switch, in the transceivers) or are modeling the breakout part of the cable itself.
